01-supp-factors.Rmd

```diff
@@ -0,0 +1,189 @@
+unordered and are an important class for statistical analysis and for plotting.
+
+Factors are stored as integers, and have labels associated with these unique
+integers. While factors look (and often behave) like character vectors, they are
+actually integers under the hood, and you need to be careful when treating them
+like strings.
+
+Once created, factors can only contain a pre-defined set values, known as
+*levels*. By default, R always sorts *levels* in alphabetical order. For
+instance, if you have a factor with 2 levels:
+
+## Tip {.callout}
+> The `factor()` command is used to create and modify factors in R
+
+```{r intro_to_factors}
+sex <- factor(c("male", "female", "female", "male"))
+```
+
+R will assign `1` to the level `"female"` and `2` to the level `"male"` (because
+`f` comes before `m`, even though the first element in this vector is
+`"male"`). You can check this by using the function `levels()`, and check the
+number of levels using `nlevels()`:
+
+```{r examining_factors}
+levels(sex)
+nlevels(sex)
+```
+
+Sometimes, the order of the factors does not matter, other times you might want
+to specify the order because it is meaningful (e.g., "low", "medium", "high") or
+it is required by particular type of analysis. Additionally, specifying the
+order of the levels allows us to compare levels:
+
+```{r, error=TRUE}
+food <- factor(c("low", "high", "medium", "high", "low", "medium", "high"))
+levels(food)
+food <- factor(food, levels=c("low", "medium", "high"))
+levels(food)
+min(food) ## doesn't work
+food <- factor(food, levels=c("low", "medium", "high"), ordered=TRUE)
+levels(food)
+min(food) ## works!
+```
+
+In R's memory, these factors are represented by numbers (1, 2, 3). They are
+better than using simple integer labels because factors are self describing:
+`"low"`, `"medium"`, and `"high"`" is more descriptive than `1`, `2`, `3`. Which
+is low?  You wouldn't be able to tell with just integer data. Factors have this
+information built in. It is particularly helpful when there are many levels
+(like the subjects in our example data set).
+
+## Challenge - Representing data in R {.challenge}
+>
+> You have a vector representing levels of exercise undertaken by 5 subjects
+>
+> **"l","n","n","i","l"** ; n=none, l=light, i=intense
+>
+> What is the best way to represent this in R?
+>
+> a) exercise<-c("l","n","n","i","l")
+>
+> b) exercise<-factor(c("l","n","n","i","l"), ordered=TRUE)
+>
+> c) exercise<-factor(c("l","n","n","i","l"), levels=c("n","l","i"), ordered=FALSE)
+>
+> d) exercise<-factor(c("l","n","n","i","l"), levels=c("n","l","i"), ordered=TRUE)
+
+### Converting factors
+
+Converting from a factor to a number can cause problems:
+
+```{r converting_factors}
+f<-factor(c(3.4, 1.2, 5))
+as.numeric(f)
+```
+
+This does not behave as expected (and there is no warning).
+
+The recommended way is to use the integer vector to index the factor levels:
+
+```{r converting_factors_correctly}
+levels(f)[f]
```

Comment:
@naupaka, I don't think this isn't a fair comparison of the clunkiness of the two methods. `levels(f)[f]` still returns a character vector. To obtain a numeric vector using this method, you'd have to run the following: `as.numeric(levels(f)[f])`. I don't see much difference in the clunkiness. Personally, I prefer `as.numeric(as.character(f))` because I can always remember it when I need to convert a factor to a number. The version with `levels` is recommended only because of a speed gain in specific use cases (see [here][so1] and [here][so2]).

@tomwright01, it's fine if you use the `levels` method, but please add `as.numeric` since the goal of this section was to convert factors to numbers.

[so1]: http://stackoverflow.com/questions/3418128/how-to-convert-a-factor-to-an-integer-numeric-without-a-loss-of-information
[so2]: http://stackoverflow.com/questions/6979625/arithmetic-operations-on-r-factors/6980780#6980780